python/ray/tune/api_v2/out_of_band_serialize_dataset.py

```diff
@@ -0,0 +1,34 @@
+import contextlib
```

Comment:
Consider putting these things in `impl/` sub-dirs.

Reply:
Done!

Reply:
> Per @richardliaw 's suggestion, could we split this PR into (1) interface and (2) implementation?
> 
> The interface PR should just be a couple of files with implementation stubbed out (raise NotImplementedError).

Hmmm, yeah I get what you are coming from.
I can see it's beneficial for:
1. When an interface can have several implementations underneath.
2. Special case to 1, where we want to have a dummy impl for unit test purposes.
3. For coordinating and collaboration when concurrent progress needs to be made for various components.

In this Tuner case, I don't think 1 has merit as we only have one real impl. 2 may be relevant especially since Trainer.fit() is calling into Tuner - this could make unit testing Trainer easier. 3. I think we have already passed that stage.

Unique to this Tuner case, is that we have Tuner wraps TunerInternal differently in ray client v.s. non ray client mode. I am concerned that adding another layer and making it Tuner, TunerImpl and TunerInternal may be a little too much.
Could you take a look at the current status of code (as I made quite some updates already, including hiding internal impl details from external users per your suggestion)? If you still think it is better to have this interface abstraction, I am happy to do so!

Reply:
The merit of splitting interface / implementation here is mainly on the reviewer focus (so, a special case for 3).

In particular, I think Eric's proposal is first to merge in a stub `tuner.py:Tuner` so that we can review/converge on that first, since there are some small minor deviations that are already occuring (RunConfig, etc). After merging those interfaces, we can just follow up with (this PR) to actually fill out the stubs.

Reply:
In particular, I would propose opening a new PR that merges in stubs + docstrings, and then rebasing this current PR on that.